pkg/api/types.go

```diff
@@ -688,36 +691,66 @@ type ServiceStatus struct{}
 
 // ServiceSpec describes the attributes that a user creates on a service
 type ServiceSpec struct {
-	// Port is the TCP or UDP port that will be made available to each pod for connecting to the pods
-	// proxied by this service.
-	Port int `json:"port"`
+	// Required: The list of ports that are exposed.
+	// FIXME: this would be nicer as a map of name -> struct
+	Ports []ServicePort
 
-	// Optional: Supports "TCP" and "UDP".  Defaults to "TCP".
-	Protocol Protocol `json:"protocol,omitempty"`
+	// Either 'Selector' or 'Endpoints' must be set:
 
-	// This service will route traffic to pods having labels matching this selector. If empty or not present,
-	// the service is assumed to have endpoints set by an external process and Kubernetes will not modify
-	// those endpoints.
+	// This service will route traffic to pods having labels matching this
+	// selector.
 	Selector map[string]string `json:"selector"`
 
+	// Use the specified endpoints.
+	// FIXME: Some weird interaction.  All the fields of Ports[] are basically
+	//   ignored in favor of Endpoints.  Maybe make this just a list of IPs,
+	//   and disallow named DestinationPorts if using this mode?
+	Endpoints []Endpoint
+
 	// PortalIP is usually assigned by the master.  If specified by the user
 	// we will try to respect it or else fail the request.  This field can
 	// not be changed by updates.
 	PortalIP string `json:"portalIP,omitempty"`
 
-	// CreateExternalLoadBalancer indicates whether a load balancer should be created for this service.
+	//FIXME: Should these next 3 be per-port (different external IPs?) or for all ports?
+	// CreateExternalLoadBalancer indicates whether a load balancer should
+	// be created for this service.
 	CreateExternalLoadBalancer bool `json:"createExternalLoadBalancer,omitempty"`
+
 	// PublicIPs are used by external load balancers.
 	PublicIPs []string `json:"publicIPs,omitempty"`
```

Comment:
Should public be set into status rather than spec?  So the target user of services cannot set publicip, but can read them?  Setting a public ip does not seem like an end app author action (an app admin or infra admin role, yes).


> On Feb 13, 2015, at 10:25 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/types.go:
> 
> >  	CreateExternalLoadBalancer bool `json:"createExternalLoadBalancer,omitempty"`
> > +
> >  	// PublicIPs are used by external load balancers.
> >  	PublicIPs []string `json:"publicIPs,omitempty"`
> Any ideas on what we can do to clean up or clarify the PortalIP vs. PublicIPs confusion? InternalIP and ExternalIPs?
> 
> —
> Reply to this email directly or view it on GitHub.
>